### ISSUE TITLE
sparse attn setting: ('full', 'sparse') required?

### ISSUE THREAD

**Issue by richcmwang** [lucidrains/DALLE-pytorch#294]
Hi @afiaka87 The sparse attn page suggests setting ('full', 'sparse') to cycle between full and sparse attention:
  
```dalle = DALLE(
    # ...
    attn_types = ('full', 'sparse')  # cycles between full and sparse attention
```
We can run a bigger model if we set `attn_types = 'sparse'` (or equivalently ('sparse', 'sparse')). Would this setting cause any issue? Thanks.

**Comment by robvanvolt** [comment]
In theory, the model "might" miss some features, but in the CogView paper the authors state that both sparse and full attention yield similar results for image generation - only sparse being more memory efficient and being faster. The authors decided to train their model with full attention, because they were afraid that their additional supersampling features would perform worse with sparse attention.

**Comment by afiaka87** [comment]
@richcmwang

So the `sparse` flag actually uses a specific type of sparse attention from Microsoft referred to as `VariableSparsity`. It operates over 16x16 blocks and it will choose a mildly different sparsity structure each time its created up to `depth` number of layers. 

While I don't think the _why_ has been consistently proven - I'm seeing it referenced in many papers that using a fully sparse transformer is the way to go and actually helps with performance and loss.

OpenAI did this on the images for DALL-E; axial column/axial row attention are a type of sparse attention which is designed to focus on the 2-dimensional nature of the dataset (keep in mind - in DALL-E the text receives full causal attention no matter what; sparse attention is only applied to the image.

There is an interesting relationship here - and I think it has to do with modality. Unfortunately I know little else. I believe this research is out there but if anyone could cite a paper that would be great as I'd love to understand the relationship between sparse structures/attention and dealing with different modalities. 

Unrelated to a degree, but while you're tuning things;
Perhaps the single most relevant paper to "which parameters should I use" is the "Scaling Laws" (for transformers) paper which is specifically about the impacts of various hyperparameters and which ones to ignore/focus on. In particular, they find that you should effectively try to scale the total number of parameters to a point as high as will work with your compute constraints (even if this means doing things you may consider strange like decreasing depth to increase to total dimension of the transformer).  There is a sweet spot of parameter size where, despite the fact that this thing is humongous and only runs at something very slow like say, 4 samples per second, the size of the model will _still_ improve the loss in the same compute timespan because each given step in training is that much more effective.

Finding this sweet spot is tough; but using a fully sparse transformer with Microsoft's sparse attention is definitely one way to go about it.

**Comment by afiaka87** [comment]
Another thing to consider - the loss curve. 

The lowest amount of loss possible on your dataset will never change. Your dataset, however, can. So will loss curves be different for different datasets? 

Yes. The loss can actually only go to as low as the "average entropy" of your dataset. This is well-discussed in the literature for training with language - not so much with vision. 

In the vision case we've already seen this play out _a lot_. Different datasets give different loss curves and it's empirically related to how consistent your captions are, how clean they, and definitely how much they relate to "everything in the picture".

For instance - the mannequin in my profile was generated from a DALLE I trained on the images released in the OpenAI blogpost for DALLE. Its loss goes to 1 and immediately starts to overfit. This is because there's a white background, the subject in the image is consistently centered in the same place, and the captions refer essentially _only_ to what's in the image e.g. "a male mannequin wearing a green checkered t-shirt and brown pleated pants". 

What's important to note here is that the loss curve in this instance is (almost) _completely_ unrelated to the parameters of the model and certainly to the size of the dataset which would give the same loss even if I only used half as much data. 

The scale of a dataset only matters when the base information _in_ that dataset has a higher amount of natural entropy. When dealing with real photos with no "rules" about how they're taken - the entropy goes up and the ability to generalize requires more data. Same with your captions. How much can you possibly glean about an image from the reddit comment _about_ that image? How does the token "lol memes" relate to a picture of Barack Obama? The answer? It kind of doesnt and is a fleeting cultural knowledge which isn't a universal truth. This is entropy.
